packages/sanity/test/descriptors/utils.ts

```diff
@@ -0,0 +1,218 @@
+/* eslint-disable complexity */
+import {cloneDeep} from 'lodash'
+
+function deleteKeyDeep(obj: any, target: string) {
+  if (!obj || typeof obj !== 'object') {
+    return
+  }
+
+  delete obj[target]
+
+  for (const key of Object.keys(obj)) {
+    if (Array.isArray(obj[key])) {
+      for (let i = 0; i < obj[key].length; i++) {
+        deleteKeyDeep(obj[key][i], target)
+      }
+    } else if (typeof obj[key] === 'object') {
+      deleteKeyDeep(obj[key], target)
+    }
+  }
+}
+
+// Check if options contain unserializable types
+function deleteUnserializableTypes(obj: any): any {
+  if (!obj || typeof obj !== 'object') return obj
+
+  const o = cloneDeep(obj)
+
+  // Handle arrays
+  if (Array.isArray(o)) {
+    const filtered = o
+      .map((item) => deleteUnserializableTypes(item))
+      .filter((item) => item !== undefined)
+
+    return filtered.length > 0 ? filtered : undefined
+  }
+
+  // Handle objects
+  for (const key of Object.keys(o)) {
+    const value = o[key]
+
+    if (value === null) {
+      // Keep null values if you want, or delete them
+      continue
+    }
+
+    if (typeof value === 'object') {
+      // Check for unserializable types
+      if (value.__type === 'function' || value.__type === 'cyclic' || value.__type === 'maxDepth') {
+        delete o[key]
+      } else {
+        // Recursively clean nested objects/arrays
+        const cleaned = deleteUnserializableTypes(value)
+
+        if (cleaned === undefined) {
+          delete o[key]
+        } else if (
+          typeof cleaned === 'object' &&
+          !Array.isArray(cleaned) &&
+          Object.keys(cleaned).length === 0
+        ) {
+          // Delete empty objects
+          delete o[key]
+        } else {
+          o[key] = cleaned
+        }
+      }
+    }
+  }
+
+  // Return undefined if the object is now empty
+  if (Object.keys(o).length === 0) {
+    return undefined
+  }
+
+  return o
+}
+
+function onlyNoopFieldsets(fieldsets: any[], groups: any[]) {
+  // extractWorkspaceManifest excludes fieldsets which are singluar (.single == true) except for when there are groups
+  if (fieldsets.length === 1 && (groups || []).length <= 1) {
+    return true
+  }
+
+  return fieldsets.every(
+    (fieldset) =>
+      Object.values(fieldset).filter((v) => {
+        return v !== undefined
+      }).length <= 1,
+  )
+}
+
+export function normalizeConvertedType(obj: any) {
+  if (!obj) return obj
```

Comment:
Don't we need this function in @sanity/schema/_internal (e.g. `normalizeFromManifest`) if we're going to use it elsewhere?

Reply:
Also in `@sanity/schema` we have the right type of this.